hudi-common/src/test/java/org/apache/hudi/avro/TestHoodieAvroUtils.java

```diff
@@ -1231,4 +1234,458 @@ void recordNeedsRewriteForExtendedAvroTypePromotion(Schema writerSchema, Schema
+        }
+        // Choose a non-null type from the union
+        Schema nonNullType = types.stream()
+            .filter(t -> t.getType() != Schema.Type.NULL)
+            .findFirst()
+            .orElse(types.get(0));
+        return generateRandomValue(nonNullType, null);
+      }
+      // Handle default values
+      if (defaultValue != null
+          && !(defaultValue instanceof org.apache.avro.JsonProperties.Null)
+          && RANDOM.nextBoolean()) {
+        return defaultValue;
+      }
+      // For nullable fields, sometimes return null
+      if (schema.getType() == Schema.Type.UNION && schema.getTypes().size() == 2
+          && schema.getTypes().get(0).getType() == Schema.Type.NULL && RANDOM.nextBoolean()) {
+        return null;
+      }
+      Schema actualSchema = schema;
+      try {
+        actualSchema = resolveNullableSchema(schema);
+      } catch (Exception e) {
+        // If we can't resolve the schema, just use the original
+        actualSchema = schema;
+      }
+
+      switch (actualSchema.getType()) {
+        case NULL:
+          return null;
+        case BOOLEAN:
+          return RANDOM.nextBoolean();
+        case INT:
+          return RANDOM.nextInt(1000);
+        case LONG:
+          return RANDOM.nextLong() % 1000000L;
+        case FLOAT:
+          return RANDOM.nextFloat() * 100f;
+        case DOUBLE:
+          return RANDOM.nextDouble() * 1000.0;
+        case STRING:
+          return "test_string_" + RANDOM.nextInt(1000);
+        case BYTES:
+          byte[] bytes = new byte[RANDOM.nextInt(10) + 1];
+          RANDOM.nextBytes(bytes);
+          return ByteBuffer.wrap(bytes);
+        case RECORD:
+          return generateRandomRecord(actualSchema);
+        case ENUM:
+          List<String> symbols = actualSchema.getEnumSymbols();
+          return new GenericData.EnumSymbol(actualSchema, symbols.get(RANDOM.nextInt(symbols.size())));
+        case ARRAY:
+          List<Object> array = new ArrayList<>();
+          int arraySize = RANDOM.nextInt(3) + 1;
+          for (int i = 0; i < arraySize; i++) {
+            array.add(generateRandomValue(actualSchema.getElementType(), null));
+          }
+          return array;
+        case MAP:
+          Map<String, Object> map = new HashMap<>();
+          int mapSize = RANDOM.nextInt(3) + 1;
+          for (int i = 0; i < mapSize; i++) {
+            map.put("key_" + i, generateRandomValue(actualSchema.getValueType(), null));
+          }
+          return map;
+        case FIXED:
+          byte[] fixedBytes = new byte[actualSchema.getFixedSize()];
+          RANDOM.nextBytes(fixedBytes);
+          return new GenericData.Fixed(actualSchema, fixedBytes);
+        default:
+          return null;
+      }
+    }
+
+    /**
+     * Compare a SpecificRecord with a GenericRecord using the same schema
+     */
+    public static void assertRecordsEqual(SpecificRecord specificRecord, GenericRecord genericRecord, Schema schema) {
```

Comment:
Can we leverage the `deepCopy` functionality to copy the generic record into a new specific record? Then we would not need the custom equals implementation here.